src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -46,43 +55,144 @@ class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig)(implicit v
+
+  override def deleteCluster(googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
+    //currently, the bucketPath of the clusterRequest are not used - it will be used later as a place to store notebooks and results
+    val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultRegion, clusterName)
+    executeGoogleRequestAsync(googleProject, clusterName)(request).mapTo[Unit] recover {
+      // Treat 404 errors as a successfully deleted cluster
+      case CallToGoogleApiFailedException(_, _, 404, _) => ()
     }
   }
 
+  override def getClusterStatus(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[LeoClusterStatus] = {
+    getCluster(googleProject, clusterName).map { cluster =>
+      LeoClusterStatus.withNameIgnoreCase(cluster.getStatus.getState)
+    }
+  }
 
-  def deleteCluster(googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Unit] = {
-    Future {
-      //currently, the bucketPath of the clusterRequest are not used - it will be used later as a place to store notebooks and results
-      val request = dataproc.projects().regions().clusters().delete(googleProject, dataprocConfig.dataprocDefaultZone, clusterName)
-      try {
-        executeGoogleRequest(request)
-      } catch {
-        case e:GoogleJsonResponseException =>
-          if(e.getStatusCode!=404)
-            throw CallToGoogleApiFailedException(googleProject, clusterName, e.getStatusCode, e.getDetails.getMessage)
+  override def getClusterMasterInstanceIp(googleProject: GoogleProject, clusterName: String)(implicit executionContext: ExecutionContext): Future[Option[String]] = {
+    // OptionT is handy when you potentially want to deal with Future[A], Option[A],
+    // Future[Option[A]], and A all in the same flatMap!
+    val ipOpt: OptionT[Future, String] = for {
+      cluster <- OptionT.liftF[Future, Cluster] { getCluster(googleProject, clusterName) }
+      masterInstanceName <- OptionT.fromOption { getMasterInstanceName(cluster) }
+      masterInstanceZone <- OptionT.fromOption { getZone(cluster) }
+      masterInstance <- OptionT.liftF[Future, Instance] { getInstance(googleProject, masterInstanceZone, masterInstanceName) }
+      masterInstanceIp <- OptionT.fromOption { getInstanceIP(masterInstance) }
+    } yield masterInstanceIp
+
+    // OptionT[Future, String] is simply a case class wrapper for Future[Option[String]].
+    // So this just grabs the inner value and returns it.
+    ipOpt.value
+  }
+
+  override def getClusterErrorDetails(operationName: String)(implicit executionContext: ExecutionContext): Future[Option[ClusterErrorDetails]] = {
+    val errorOpt: OptionT[Future, ClusterErrorDetails] = for {
+      operation <- OptionT.liftF[Future, Operation] { getOperation(operationName) }
+      _ <- OptionT.pure { operation.getDone }.filter(_ == true)
```

Comment:
Maybe a few in line comments here in plain english about what  `liftF` and `pure` do might be good here

Reply:
Sure. They're just different ways of creating an `OptionT`. Basically:
- `OptionT.pure` turns an `A` into an `OptionT[F, A]`
- `OptionT.liftF` turns an `F[A]` into an `OptionT[F, A]` 
- `OptionT.fromOption` turns an `Option[A]` into an `OptionT[F, A]`